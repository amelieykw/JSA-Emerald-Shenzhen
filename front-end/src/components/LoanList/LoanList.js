import React from 'react'
import { Container, Content, Text } from 'native-base'

import style from './LoanListStyle'
import data from '../../../helpers/fakeLoanData'

import LoadCard from '../LoanCard/LoanCard'
import NewLoanButton from '../NewLoanButton/NewLoanButton'

function LoanList() {
  return (
    <Container>
      <Content style={style.loanHomeScroll}>
        <Text style={style.header}>Current active contracts</Text>
        {data.loans.map(loan => {
          return <LoadCard key={loan.id} loan={loan} />
        })}
      </Content>
      <NewLoanButton />
    </Container>
  )
}

LoanList.navigationOptions = {
  title: 'Loans'
}

export default LoanList
