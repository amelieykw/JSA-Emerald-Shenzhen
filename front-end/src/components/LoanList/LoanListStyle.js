import { StyleSheet } from 'react-native'

const style = StyleSheet.create({
  loanHomeScroll: {
    paddingHorizontal: 10,
    flex: 1
  },
  navigator: {
    position: 'absolute',
    width: '106%',
    backgroundColor: '#2296F3',
    height: 75,
    display: 'flex',
    flexDirection: 'row',
    alignItems: 'center',
    paddingTop: 20,
    paddingLeft: 20,
    shadowColor: '#000',
    shadowOffset: {
      width: 0,
      height: 1
    },
    shadowOpacity: 0.34,
    shadowRadius: 6.27,

    elevation: 5
  },
  navigatorText: {
    color: '#fff',
    paddingLeft: 30,
    fontSize: 20
  },
  header: {
    fontSize: 24,
    color: '#444444',
    marginVertical: 15
  }
})

export default style
