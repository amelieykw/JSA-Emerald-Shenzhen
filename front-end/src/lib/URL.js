import { BASE_URL as BASE } from 'react-native-dotenv'

const BASE_URL = BASE || 'http://10.22.18.66:3000'

const URL = {
  getLoans: `${BASE_URL}/loans`
}

export default URL
