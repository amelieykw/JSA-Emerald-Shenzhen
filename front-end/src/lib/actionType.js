const ACTION_TYPE = {
  INIT_LOANLIST: 'INIT_LOANLIST',
  INIT_LOANLIST_SUCCESS: 'INIT_LOANLIST_SUCCESS',
  INIT_PRODUCTLIST: 'INIT_PRODUCTLIST',
  INIT_PRODUCTLIST_SUCCESS: 'INIT_PRODUCTLIST_SUCCESS'
}

export default ACTION_TYPE
