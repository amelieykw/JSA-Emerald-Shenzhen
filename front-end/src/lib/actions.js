import URL from './URL'

const ACTION_TYPE = {
  LOADING: 'LOADING',
  LOADDONE: 'LOADDONE',
  INIT_LOANLIST: 'INIT_LOANLIST',
  INIT_LOANLIST_SUCCESS: 'INIT_LOANLIST_SUCCESS',
  INIT_PRODUCTLIST: 'INIT_PRODUCTLIST',
  INIT_PRODUCTLIST_SUCCESS: 'INIT_PRODUCTLIST_SUCCESS'
}

function fetchLoanList() {
  return async dispatch => {
    dispatch({ type: ACTION_TYPE.LOADING })
    dispatch({ type: ACTION_TYPE.INIT_LOANLIST })

    try {
      const data = await fetch(URL.getLoans, {
        headers: { Authentication: 'Bearer token', Accept: 'application/json' }
      }).then(res => res.json())

      dispatch({ type: ACTION_TYPE.INIT_LOANLIST_SUCCESS, loanList: data })
      dispatch({ type: ACTION_TYPE.LOADDONE })
    } catch (error) {
      // haven`t handled
      dispatch({ type: ACTION_TYPE.ERROR, error })
    }
  }
}

export { fetchLoanList, ACTION_TYPE }
