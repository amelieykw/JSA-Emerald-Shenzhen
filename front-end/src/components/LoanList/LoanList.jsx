import React, { useContext, useEffect } from 'react'
import { Content, Text, Container, Spinner } from 'native-base'
import { NavigationContext } from 'react-navigation'
import PropTypes from 'prop-types'
import { connect } from 'react-redux'
import style from './LoanListStyle'
import LoadCard from './LoanCard/LoanCard'
import { fetchLoanList } from '../../lib/actions'
import FloatingButton from '../FloatingButton/FloatingButton'

function LoanList({ loanList, loading, fetchData }) {
  const navigation = useContext(NavigationContext)
  const handlePress = () => {
    navigation.navigate('NewLoan')
  }

  useEffect(() => {
    fetchData()
  }, [])
  return (
    <Container>
      {loading ? (
        <Spinner color="blue" />
      ) : (
        <>
          <Content style={style.loanHomeScroll}>
            <Text style={style.header}>Current active contracts</Text>
            {loanList.map(loan => {
              return <LoadCard key={loan.id} loan={loan} />
            })}
          </Content>
          <FloatingButton
            icon={{ type: 'Entypo', name: 'plus' }}
            text="New Loan"
            handlePress={handlePress}
            buttonStyle={{ width: 160 }}
          />
        </>
      )}
    </Container>
  )
}

const mapStateToProps = state => {
  return { loanList: state.loanList, loading: state.loading }
}
const mapDispatchToProps = dispatch => {
  return {
    fetchData: () => dispatch(fetchLoanList())
  }
}

LoanList.navigationOptions = {
  title: 'Loans'
}
LoanList.propTypes = {
  loanList: PropTypes.arrayOf(PropTypes.any).isRequired,
  fetchData: PropTypes.func.isRequired,
  loading: PropTypes.bool.isRequired
}

export default connect(mapStateToProps, mapDispatchToProps)(LoanList)
