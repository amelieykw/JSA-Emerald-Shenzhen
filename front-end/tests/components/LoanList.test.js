import React from 'react'
import renderer from 'react-test-renderer'
import configureStore from 'redux-mock-store'
import thunk from 'redux-thunk'
import { Provider } from 'react-redux'
import LoanList from '../../src/components/LoanList/LoanList'
import data from '../../helpers/mockData_FE'
import { fetchLoanList, ACTION_TYPE } from '../../src/lib/actions'

const mockStore = configureStore([thunk])

describe('<LoanList />', () => {
  let store
  let component
  let payloads
  beforeEach(() => {
    store = mockStore({ loans: { loanList: data.loans }, loading: true })

    payloads = [
      {
        type: ACTION_TYPE.LOADING
      },
      {
        type: ACTION_TYPE.INIT_LOANLIST
      },
      {
        type: ACTION_TYPE.INIT_LOANLIST_SUCCESS,
        loanList: {}
      },
      {
        type: ACTION_TYPE.LOADDONE
      }
    ]
  })

  it('<LoanList /> renders correctly', () => {
    component = renderer.create(
      <Provider store={store}>
        <LoanList />
      </Provider>
    )
    expect(component.toJSON()).toMatchSnapshot()
  })

  it('<LoanList /> has 2 child', () => {
    expect(component.toJSON().children.length).toBe(1)
  })

  it('<LoanList /> should dispatch an action when mounted', async () => {
    renderer.act(() => {
      component = renderer.create(
        <Provider store={store}>
          <LoanList />
        </Provider>
      )
    })

    // eslint-disable-next-line no-undef
    window.fetch = jest.fn().mockImplementation(() =>
      Promise.resolve({
        json: () => Promise.resolve({})
      })
    )

    await store.dispatch(fetchLoanList())
    const actions = store.getActions()

    expect(actions.pop()).toEqual(payloads.pop())
    expect(actions.pop()).toEqual(payloads.pop())
    expect(actions.pop()).toEqual(payloads.pop())
    expect(actions.pop()).toEqual(payloads.pop())
  })
})
