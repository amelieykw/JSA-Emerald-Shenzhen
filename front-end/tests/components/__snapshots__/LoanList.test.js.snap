// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<LoanList /> <LoanList /> renders correctly 1`] = `
<View
  style={
    Object {
      "backgroundColor": "#fff",
      "flex": 1,
      "height": 1334,
    }
  }
>
  <ActivityIndicator
    animating={true}
    color="blue"
    hidesWhenStopped={true}
    size="large"
    style={
      Object {
        "height": 80,
      }
    }
  />
</View>
`;
